### ISSUE TITLE
Performance optimizations for WorldObject transforms

### ISSUE THREAD

**Issue by FirefoxMetzger** [pygfx/pygfx#520]
#482 refactored all the transformations to use numpy and pylinalg. The PR did not explicitly tweak performance beyond what is naturally gained/lost by switching to numpy, because the refactor itself was already humongous enough to be challenging to review. All optimization opportunities will/shall be done in future PRs, and this issue lists a few of them.

> **Note**: I am aware that this issue is well within TL:DR territory and I'm sorry that this turned into a massive wall of text. However, I wanted to have the full picture in one place, and some of these ideas are a bit more complex and need more words to get across.
>
> To make this less intimidating, I added proper sectioning and I strongly encourage you to read the section headings first to get an idea of what information is where, and to then decide which parts are worth reading :)

## General Benchmarking Note

> **Note**: One thing that is important to highlight before going into optimization ideas is that we currently don't have good benchmarking scenarios to judge the impact of optimizations. Our example scenes do most of their work in `before_render`, but our typical use cases do most of the work during keyboard/mouse events. 

The above note is important, because `before_render` runs in-sync with a draw call whereas other events are asynchronously. As a result, doing more work in total, but less work in-sync with a draw is a desirable tradeoff, because faster draw calls mean more FPS and a smoother render. Unfortunately, we can't really measure this tradeoff, so putting some thought into benchmark scenarios will give us a lot of clarity moving forward.

The new transform API makes this tradeoff by eagerly syncing WorldObject properties and the world transform. As a result, all data is up-to-date and neatly staged once the draw call starts. We only need to send updates to the GPU and process them. At the same time, this may mean more work per event callback, because calling a setter of a transform property implies an update of the world transform.

## Mental model of the transform API

The second preliminary is a mental model of the transform API. This isn't super relevant to users that consume pygfx (they don't care), but it is useful for devs of pygfx, since it suggests optimizations and places to look when hunting bugs.

**The canonical source of truth**

The biggest change from the old system is that the transform API chooses _exactly one_ explicit source of truth for its transformations; everything else is a computed property. At the moment, this source is `AffineTransform.untracked_matrix` aka `obj.local.untracked_matrix`. This means that everything else that can be accessed via `obj.local.xxx` or `obj.world.xxx` is a property derived from (one or more) `obj.local.untracked_matrix` values. Setters of properties work the same way: Once we get a new value for a property, we work out the corresponding value for `obj.local.untracked_matrix` and update `untracked_matrix`. 

One consequence of this implementation is that `AffineTransform` doesn't require exclusive ownership of `untracked_matrix`. External updates of `untracked_matrix` never put `AffineTransform` into an invalid/inconsistent state and to allow exploiting this `AffineTransform` only ever modifies `untracked_matrix` in place. This means that `AffineTransform` is essentially a (fancy) `array.view()` into `untracked_matrix` with lots of useful properties. The same is true for `RecursiveTransform` (the class that backs `obj.world.xxx`) which is completely stateless except for `reference_up` and some other metadata that affect the value of its properties (but not the transform itself).

This is awesome (I think) because it decouples the WorldObject's transform API and the underlying data model. At the moment each transform uses a "fresh" and independent numpy array, but nothing stops us from using a view or other structure instead. For example, we can make `AffineTransform` wrap the affine matrix from a uniform buffer and directly get/set its properties. Alternatively, we can pre-init a large batch of matrices and wrap views into single batch elements with `AffineTransform`. This way, we can create and remove WorldObjects without causing memory fragmentation. If we are smart when placing the transforms, we can even enable bulk updates because we can store children contiguously in memory. In sum, we can be very flexible in how we optimize while keeping the API as-is (and I find this decoupling quite neat).

Another consequence of the choice of having a canonical source of truth is that we get lazy evaluation for free across the entire scene graph and defer computing any value until we actually need it. Thus, updating a transform is incredibly cheap (replace the values, done). At the same time, reading a property can become expensive. For example `var = obj.world.position` now needs (1) matmul all the parent matrices until we reach the world frame, (2) decompose the resulting matrix, (3) return the requested position. 

**Caching of properties**

To reduce the (potentially large) amount of computing that happens when a property is accessed we cache property values. Caches are invalidated whenever a `untracked_matrix` on which the property depends updates. We also cache (and reuse) intermediate results whenever possible to further reduce the compute needed after a matrix has changed. This adds more state to our (otherwise lean) `AffineTransform` and `RecursiveTransform`, but this state is "non-critical", i.e., were we to pickle/serialize a transform, we would discard the caches without losing information.

Unfortunately, it is impossible to directly monitor a numpy array for changes. The reason is that we can't (cheaply) detect in-place updates of an array or its slices and modifications can happen through various channels including low-level numpy kernels which don't use `__setitem__` and friends. Instead, we track changes using a clock/timestamp which is stored in `AffineTransform.last_modified`. Caches compare to this timestamp to decide if they are still valid (cache time > last_modified) or stale. This shifts the workload towards the getters and keeps updates cheap (replace values, update one counter, done).

A consequence of this inability to track changes is that in-place updates of `untracked_matrix` from external sources are not tracked (hence the name). They need to manually invalidate the property caches to make this optimization work correctly. This is easy to do by calling `obj.local.flag_update()` or `obj.world.flag_update()` respectively.

**Eager/Synchronous Updates**

Computing everything lazily from `obj.local.untracked_matrix` works great, but it does mean that `obj.world.matrix` is a computed property, too. By itself this is not a problem, however, WorldObjects now have two places where they store the world matrix: `obj.world.matrix` and `obj.uniform_buffer.data["transform_matrix"]`. These need to be kept in sync, otherwise, we will get wrong (and typically weird looking) renders.

Previously, we did this in `renderer.render(..)` by walking over the entire scene graph and updating `uniform_buffer`s where needed. Theoretically, we could keep doing this, but we dislike workload during draw. Worse, since all computed properties are evaluated lazily, we effectively defer all world matrix updates until the next draw, which means lots of matmul when we least want it.

Instead, `AffineTransform` (and `RecursiveTransform`) feature a callback mechanism via `on_update` that allows subscribing to their respective `flag_update()` events. This effectively allows external parties to update themself in-sync with the transform to which they subscribe.

Using this, we can once again shuffle the workload away from the draw call by synchronizing the uniform buffer whenever `ob.world` invalidates its caches. This happens whenever `obj.local` or any of the object's parent transformations updates. It significantly increases the workload done during an update (we copy the world transforms of all children on each update), but is the price we pay for keeping the draw event lean.

This callback system is also currently misused by the lights and helpers (I think this is all, but not sure), which update their state whenever the tracked transform changes. This is a relic of the old system though and should be refactored (it has nothing to do with the transform API).

## Big Optimization Opportunities

With the theory out of the way, here are ways to further improve the engine. The two biggest slowdowns right now stem from (1) excessive array creation, and (2) the underlying (informal) data model. There are some other improvements and refactors that seem useful, but they are fairly use-case specific and would need a benchmark suite to prioritize/decide on.

**Excessive Array Creation**

At the moment, calls to pylinalg as well as all computed properties will return a new/fresh array. The only exception are cache hits of properties, which will return the previously cached array. This results in a lot of allocating and deallocating of numpy arrays, which is much slower than normal object creation, because we cross the Python-C-boundary a lot. To improve this, we need the following:

- [ ] Smarten up the property cache by (1) giving it a reusable buffer and (2) exposing this buffer to the compute functions
- [ ] Rely more on pylinalgs in-place operations (`out=...`)
- [ ] Optionally: Allow caches to wrap existing buffers (to get rid of the duplicate world matrix in WorldObject)

We may also want to pre-allocate local transform matrices as a batch and assign slices/elements to WorldObjects. This may not have a big impact now but will come in handy for scenes that do a lot of object creation and destruction, e.g., when using particle systems.

**Data Model**

The other big improvement will come from a more formal model on how pygfx manages its data.

At the moment, pygfx uses a "doubly-linked tree" structure, meaning that WorldObjects are nodes in a tree with a ref from the parent to the children and a (weakref) back (aka. doubly-linked). Nodes also have a ref to their transforms, buffers, and other data. This structure is useful for fast-changing scenes that need cheap insert and remove operations and for situations where the transforms and buffers are not homogeneous and thus can't be batched.

A better model would be to give each WorldObject a small array that it can use to buffer child transforms (`self.child_buffer = np.empty((initial_size, 4, 4), dtype=float)`). Whenever an new object is added as a child (`obj.add(new_obj`) the child's `untracked_matrix` will be switched to `self.child_buffer[idx]` where `idx` is the index of the child. This way, children have their transforms aligned in a contiguous block of memory and can be accessed as a group via

```python
child_transforms = obj.child_buffer[:len(obj.children), ...]
```

We have now effectively vectorized the child transforms, and can let numpy shine by doing things like `world_transforms = obj.world.matrix @ child_transforms` which would compute all the children's world transforms in a single function call (with SIMD acceleration, potential to multi-thread, etc.). 

We can spin this idea further (to include grand-children), but that becomes more complex, since we would have to look into array-backed trees that store elements in a mix of depth-first and breadth-first fashion, as well as potentially decomposing WorldObjects to make them columnar. This is certainly possible, and maybe desirable. However, it is quite complex and we would want to decide if that complexity is worth it first before pursuing this.

## Other Optimization Opportunities

**Optimizing world matrix updates**

Another optimization that follows a similar line of thinking as the data model is to add caches to the scene graph/WorldObject itself. 

The first idea is an (aligned) cache of the world matrix. Similar to `self.child_buffer`, we can have a `self.child_world_buffer` that is used by `obj.children[idx].world.matrix` (assuming we implement the optional point from the array creation section). We know that the world transform of all children needs updating whenever the parent (or grand-parent) updates, so we can save some overhead by batch-updating the cache via `obj.child_world_buffer[:len(obj.children)] = obj.world.matrix @ obj.child_buffer[:len(obj.children)]` and a subsequent update of the timestamp of the cache.

In a similar fashion, we can reduce the work needed to update world matrices if we know that a sub-tree of our scene graph is static. In this case, we can compute (and cache) the transforms from each node in the sub-tree to the root node of the tree. We then store these transforms on the root node (in an aligned matrix batch) and point each `(grand)child.world.matrix` cache to this aligned structure. Once the root updates we can then compute the new world matrix of all children (and grand-children) with a single `root.world.matrix @ root.cached_transforms`.

However, both of these optimizations rely on the structure of the scene graph being largely static, i.e., if WorldObjects come and go at a rapid pace, we will spend more time updating the caches than we do enjoying the speedup. As such they should probably be opt-in caches that users choose to add when they know that this assumption is true for their scene.

**Reducing compute of transform properties**

At the moment, the implementation of most transform properties is somewhat naive, This is especially true for `position`, `rotation`, and `scale` (perhaps the most important ones), which rely on a full decomposition of the matrix. We can speed this up by, for example, just extracting the position component of the matrix instead of doing a full decomp.

**Comment by Korijn** [comment]
I created https://github.com/pygfx/pylinalg/issues/71 to track a similar optimization opportunity in the pylinalg repo.

**Comment by panxinmiao** [comment]
While using the current Transform system, I encountered some serious performance issues.
See: https://github.com/pygfx/pygfx/pull/715#issuecomment-2053385803

After digging deeper, I came across this documentation that provides a detailed explanation of the design philosophy behind the current Transform system. The article is highly professional and thorough.

### Immediate reactive updates and delayed updates

The main issue I think is that the article attempts to construct an immediate reaction mechanism to ensure that the world matrix of objects is always kept up-to-date. For a rendering engine, this is highly inefficient, especially in scenarios with a large number of frequently transforming objects.

A better approach is to simply mark the world matrix as invalid (dirty data) and then perform a one-time recursive update starting from the root node of the scene before rendering. This method is more efficient and reasonable:

1. Avoids redundant recalculation of the world matrix every time a property changes, reducing unnecessary matrix operations.
    In dynamic scenes, object properties may change frequently, and immediate updates would lead to a large amount of redundant computation. With deferred updates, the calculation is performed only once before rendering, better handling such scenarios.

2. Avoids excessive scattered computation.
    With immediate updates, modifying an object's properties requires recalculating the world matrix for that object and all its child objects, leading to a large number of scattered matrix operations. Deferred updates only require a single traversal of the scene tree before rendering, avoiding the overhead of scattered computations.

3. Computation overhead is concentrated before rendering.
    The rendering process is the most computation-intensive critical path in the entire engine. Concentrating the world matrix update calculations before rendering allows better utilization of CPU resources, avoids introducing latency at other times, and can also make the rendering process itself more efficient, improving rendering performance.

4. Truly adheres to the "lazy evaluation" principle, performing calculations only when truly necessary, improving efficiency.

5. By recursively updating from the root node before rendering, only objects that actually need updating are updated. For objects and their subtrees that have not undergone property changes, the calculation can be skipped, while immediate updates often require recalculating the entire subtree, leading to unnecessary overhead.

Moreover, from first principles, it is only before the final rendering that we can be sure that the properties of all objects in the scene will no longer change, representing the final state before rendering. Updating from the root node ensures that every node in the subtree of property-changed node undergoes the matrix multiplication operation **only once**.

### A tendency towards over-design and excessive complexity

Additionally, I feel that the article exhibits a tendency towards over-design and excessive complexity.

1. Overly pursuing decoupling and flexibility of low-level data structures.

    The article mentions designing AffineTransform as a "view" of untracked_matrix to support potential future optimizations, such as batch matrices or uniform buffers. This extreme decoupling introduces considerable implementation complexity.

    The article also mentions the idea of transforming the child objects of each WorldObject into a contiguous memory block. This design clearly increases implementation and maintenance complexity, whereas a simple tree-like hierarchical structure is sufficiently efficient for most scenarios.

2. Excessive caching and invalidation mechanisms.

    The article describes numerous caching mechanisms, timestamp checks, etc., to support lazy evaluation and timely updates. This increases implementation and maintenance complexity.

3. Over-optimization considerations for different scenarios.

    For example, pre-allocating space for child object matrices or batch computing child object world matrices may only provide optimization value in extreme large-scale static scenes but introduce more complex implementations.

4. The architecture is overly theoretical, and the feasibility of some suggestions is debatable.

   The article devotes considerable space to discussing the "mental model" of the transformation API, authoritative sources of data, and other theoretical concepts. Most of the assumptions and solution choices are based on theoretical conjectures.

    1）For instance, the article designing the ".matrix" property as a "computed property" to achieve "lazy evaluation," but in reality, this is entirely uncontrollable. With such a complex architecture and data dependencies, we cannot control when the computation actually occurs, and in actual use, the "lazy evaluation" purpose is not achieved at all.

    2）Another example is the statement:

    > but we dislike workload during draw. Worse, since all computed properties are evaluated lazily, we effectively defer all world matrix updates until the next draw, which means lots of matmul when we least want it.

    However, the actual "drawing" occurs after the CPU has submitted the draw calls to the GPU and is separate from the matrix calculations before rendering. Contrary to the idea, Concentrating matrix calculations before rendering can better utilize resources and reduce latency compared to uncontrolled scattered computations.

    3） The article also mentions the approach of using pre-allocated large memory blocks to reduce memory fragmentation, which may not necessarily be more efficient than simple dynamic allocation in actual scenarios.

Overall, I believe that this tendency towards over-design and complexity, while pursuing extreme flexibility and extensibility, may introduce unnecessary complexity and maintenance costs. Attempting to establish an overly general and theoretical framework and pursuing theoretical absolute performance optimization also overlooks the practical realities of engineering practice and the "good enough" principle.

**Comment by almarklein** [comment]
What article are you referring to?

**Comment by panxinmiao** [comment]
> What article are you referring to?

Um, I'm referring to the content in this issue. 😅 

Some different perspectives I have on some of the design philosophies of the current Transform system described in this issue.

**Comment by almarklein** [comment]
I do think the transform system, with the wishes we had for it, is a very complex problem to solve. The current solution works well and has a pretty neat API.

That said, It looks like we did not look hard enough about making it lazy. I have not looked into the code closely since, but I suspect it can be made more lazy without major changes?

At the time we considered a lot of use-cases, and found it hard to come up with a solution that covers all cases well. This is why we also considered several optimizations, and wanted to create "space"  to implement these in the future. Some of these ideas may be a bit "wild", but these are I think presented as possible directions, not things we'll implement for sure.

Anyway, would be good to have a good look at the system, make it lazier and perhaps a bit simpler.

**Comment by Korijn** [comment]
> What article are you referring to?

I'm assuming chatgpt was used here :)

**Comment by panxinmiao** [comment]
> > What article are you referring to?
> 
> I'm assuming chatgpt was used here and assumes it was fed an article :) the message is clearly generated by chatgpt to a large extent

Yes, this response is based on my different perspectives and understandings of certain issues in the original text. I utilized translation tools to assist in translating and then polished the translation, as English is not my native language. Expressing my views in detailed English, especially on complex matters, can be challenging for me.
However, our discussion revolves around the issues themselves, which I believe objectively exist.

**Comment by panxinmiao** [comment]
> That said, It looks like we did not look hard enough about making it lazy. I have not looked into the code closely since, but I suspect it can be made more lazy without major changes?
> 
> Anyway, would be good to have a good look at the system, make it lazier and perhaps a bit simpler.

I have made some attempts on this, but the complex callbacks and dependencies involved made it somewhat challenging for me.

**Comment by FirefoxMetzger** [comment]
This will be a long comment, but hopefully it will add some (clarifying 😅) context.

Whichever implementation you choose for the engine, you _need_ to decide where to store application state (the pose, mesh, shader, etc., of WorldObjects) and in which format you keep that data in memory. This choice will determine the initial performance and where the inherent complexity of your system lives. This is the one "free lunch" you get (there are better and worse choices depending on your goal). Everything beyond this becomes a tradeoff between performance, complexity, and usability:

- Want an intuitive API for users? No problem. However, it will add complexity because you have to create transformation logic between the API and the application state. (This is what all the `@property`s do in the transform module.)
- Want performance at draw? No problem. However, it will add complexity because you have to create transformation logic between the performant layout of your data and application state. You also need to be smart on when to convert because it has performance implications. (That's what the `RecursiveTransform` does which is used for the world matrix today.)
- Want lazy evaluation of properties? No problem. However, it will add complexity because you need to track which properties are up-to-date and which are out-of-date and add cache/refresh logic to keep the system consistent. (This is what the `@cached` decorator does.)
- Want a low-complexity mental model and good dev experience for the core engine? Sorry, a Genie only grants 3 wishes before he gets cast off to serve another master 😅🪔. On a more serious note: You bought the above by adding complexity, so unless you find unnecessary complexity (tech debt) you can refactor for free you have to somehow compromise on performance or usability. 

☝️ This is, essentially, why the engine is written the way it is today. These are also main "nobs" you can turn to influence performance, usability, complexity, dev experience, etc. I know it doesn't answer your concrete problem @panxinmiao but hopefully it adds some context to better understand the code.

---

Now for the actual problem of lazy vs eager updating of the world matrix. The section `Eager/Synchronous Updates` above goes into exactly this. The critical section and the performance bottleneck you are observing:

> Using this, we can once again shuffle the workload away from the draw call by synchronizing the uniform buffer whenever `ob.world` invalidates its caches. This happens whenever `obj.local` or any of the object's parent transformations updates. It significantly increases the workload done during an update (we copy the world transforms of all children on each update), but is the price we pay for keeping the draw event lean.

It comes from the design decision to keep the draw call as lean as possible. Here is the reasoning: At the time I wrote the engine the target application was one where we expected many draw calls between two consecutive changes in the scene graph. Users drag a camera around a lot (cheap to update) but modifications to the scene graph come mainly from (low-frequency) "click & point" user interactions. An efficient scene graph implementation for such a system looks very different from an efficient implementation where we expect many changes between two consecutive draw calls. This wasn't a target use-case at the time (is it now?), so we optimized for the former at expense of the latter. For animations, you are in the latter case (many updates between two draws) and thus suffer low performance due to eager updates.

You can, without too much work, rewrite this logic and do it the other way around, i.e., you can do lazy updates that wait for the draw call to compute the current set of world matrices. You could also defer sending data to the GPU until the actual draw call - I suspect would also solve the performance problem. Either way, the decision is still a tradeoff between use-cases. Your scenario (many updates per draw) will become fast; however, the other scenario (many draws per update) will become slow (slower?) because of the constant overhead you'll introduce by checking that everything is up-to-date.

What constant overhead? When the draw call starts you need to work out what has changed and update world matrices. Maybe a lot has happened, maybe nothing has changed. In the case of "nothing has changed" (many draws per update), how much time do you "waste" checking what has changed just to find that nothing changed. This degrades performance for the "many draws per update" use-case that we care about (and, hence, why I chose eager evaluation).

---

> By recursively updating from the root node before rendering, only objects that actually need updating are updated. For objects and their subtrees that have not undergone property changes, the calculation can be skipped, while immediate updates often require recalculating the entire subtree, leading to unnecessary overhead.

I wanted to keep this high-level, but I will add one comment on this before we run in the wrong direction. I thought about this idea as well back in the days and concluded that it is a dead end. There is no trivial way to "skip a subtree that didn't change". If a node changes all its children must update but all its parents remain constant. To keep things brief, this is "the wrong way around" for what we need to update recursively. We have to either walk the entire scene graph on each draw call to discover changes or add a tracking layer that marks parents as dirty/modified so that we can prune sub-trees that haven't changed.

The former adds a lot of dead weight to the "many draws per update" scenario; the latter adds another layer of complexity to an engine that you already find too complicated 😅. No easy win here.

**Comment by Korijn** [comment]
I am starting to consider taking a day off from my day job to work on this. :')

**Comment by FirefoxMetzger** [comment]
Am I adding more confusion than clarity?

**Comment by almarklein** [comment]
> What constant overhead? When the draw call starts you need to work out what has changed and update world matrices. Maybe a lot has happened, maybe nothing has changed. In the case of "nothing has changed" (many draws per update), how much time do you "waste" checking what has changed just to find that nothing changed.

This is somewhat similar to checking what buffers/texture need (partial) uploading. In that case, I solved this by using a "registry"; when an object becomes dirty, it adds itself to the registry. Then at draw time, the renderer checks the registry to get a list of objects that need updating.

**Comment by FirefoxMetzger** [comment]
> This is somewhat similar to checking what buffers/texture need (partial) uploading.

Yes. One difference is that you need to add children as well, since a world matrix depends on the chain of its parents. I.e., when the parent changes all children and grand-children need updating. 

If the registry supports efficient insert, delete, and deduplication (an object may change multiple times, but you only wish to update the world matrix once) then this can be a good solution :)

**Comment by panxinmiao** [comment]
> This will be a long comment, but hopefully it will add some (clarifying 😅) context.
> 
Thank you very much for your patient explanation.

I understand that we have many trade-offs here, such as during deferred updates, determining whether a node's Transform has changed would incur additional overhead, but Eager/Synchronous Updates may lead to unnecessary redundant matrix computations.

You mentioned that in static scenes, Eager/Synchronous Updates would have an advantage, but in dynamic scenes (especially when multiple properties of multiple objects change every frame), deferred updates would have a more significant advantage.
In static scenes, the consumption of the Transform system is almost negligible (only the cost of traversing the scene graph, but in the rendering loop, traversing the scene graph is almost necessary, not just for the Transform system). However, in dynamic scenes, the performance loss of the Transform system is indeed one of the most critical elements determining the final performance of the program.

**Comment by almarklein** [comment]
> If the registry supports efficient insert, delete, and deduplication this can be a good solution :) 

Good point. The registry must maintain order, or be sort of tree-aware.

**Comment by FirefoxMetzger** [comment]
> In static scenes, the consumption of the Transform system is almost negligible (only the cost of traversing the scene graph, but in the rendering loop, traversing the scene graph is almost necessary, not just for the Transform system).

@panxinmiao Not exactly. In the eager setup we never traverse the scene graph during draw ... at least not for transforms. Since we compute eagerly, by the time we enter the draw call everything is already up-to-date and no work has to be done.

There might be other things that require us to parse the entire tree; though, ad-hoc, I can't think of anything. We could have new textures, morphing meshes, etc., but these are all changes that are local to a world object, i.e., they don't require knowledge of where an object is in the scene graph.
